### PR TITLE
Replace langdetect with fasttext (followup of #1969)

### LINKED ISSUE #2018: Users language selection should take precedence over autodetection
### requested feature

The core of SearXNG should provide the ability to distinguish between explicit language selection by the user and language suggestions from a plugin.

### background of the request

In PR-1969 we ..

- https://github.com/searxng/searxng/pull/1969

Autodetection of the language from the search term is a great feature often requested by our users.  It should become active by default but it is not enabled by default since it has one drawback --> **The autodetection takes precedence over user's selection and this can make SearXNG unusable for the user.**

----

copied from https://github.com/searxng/searxng/pull/1969#issuecomment-1345294431

SearXNG's locale of a query comes from (*highest wins*):

1. The ``Accept-Language`` header from user's HTTP client.
2. The user select a locale in the preferences.
3. The user select a locale from the menu in the query form (e.g. ``:zh-TW``)
4. Autodetection plugin is activated in the preferences and the locale (only the language code / none region code) comes from the fastText's language detection.

Conclusion: There is a conflict between the language selected by the user and the language from language detection of this plugin.  For example, the user explicitly selects the German locale via the search syntax to search for a term that is identified as an English term. 

Try ``:de-DE thermomix``, for example ..

In this screenshot the user selects the `de-DE` locale but the autodetect plugin has switched the language to `en` because the name of the product the user looks for is a English word

grafik

### PR DESCRIPTION
**Pull request by ArtikusHG**
## What does this PR do?

This PR replaces langdetect with fasttext. Since fasttext is both faster and more accurate than langdetect, and we're using fasttext in #1969 anyway, we don't need two language detection libraries.

## Why is this change important?

This change is important to make sure SearXNG doesn't use unnecessary resources and libraries, and provides accurate language detection.

## How to test this PR locally?

Just run the code with my changes.

## Author's checklist

~This PR depends on #1969.~ already merged.

~Note: I didn't do extensive testing *yet*, this is just a PoC. Nothing broke catastrophically on my side though so that's good. Suggestions and improvements are absolutely welcome.~ Now that I have access to checks, I fixed some things and they all seem to pass now.

### CODE REVIEW

**Comment by HLFH** [comment]
Can this be merged? Thanks

**Comment by ArtikusHG** [comment]
I think so, but we need to do more extensive testing to ensure nothing is broken. The code passes all tests and runs, but there could be some less obvious issues. Also, this doesn't take care of the docker image - I have little experience with docker, so the maintainers of the docker image could help a lot here.

File: requirements.txt [hunk 1]
```diff
@@ -11,7 +11,6 @@ httpx[http2]==0.21.2
 Brotli==1.0.9
 uvloop==0.17.0
 httpx-socks[asyncio]==0.7.2
-langdetect==1.0.9
```
Comment (dalf) [inline requirements.txt:14]: the PR should be rebase on the master branch: `fasttext-wheel` is missing.

File: searx/search/checker/impl.py [hunk 2]
```diff
@@ -208,14 +219,14 @@ def _record_error(self, message: str, *args) -> None:
...
-            r = detect_langs(str(text))  # pylint: disable=E1101
-        except LangDetectException:
+        r = get_model().predict(str(text), k=1, threshold=0.3)
+        if not r:
             return None
 
-        if len(r) > 0 and r[0].prob > 0.95:
-            self.languages.add(r[0].lang)
-            self.test_results.add_language(r[0].lang)
+        if len(r) > 0 and r[1][0] > 0.95:
```
Comment (dalf) [inline searx/search/checker/impl.py]: When no language is detected, the result is still a tuple.
Possible fix:
```python
if len(r) == 2 and len(r[0]) > 0 and r[1][0] > score:
```

**Review by dalf — CHANGES_REQUESTED** [review]
Thanks you for this PR.

However, the `detect_language` function should be defined in one place to avoid duplicate code. The `searx.utils` module is good fit for that.

Here a patch that does that with some tests:
```diff
diff --git a/searx/plugins/autodetect_search_language.py b/searx/plugins/autodetect_search_language.py
index 034668041..625f85373 100644
--- a/searx/plugins/autodetect_search_language.py
+++ b/searx/plugins/autodetect_search_language.py
@@ -66,46 +66,29 @@ that is identified as an English term (try ``:de-DE thermomix``, for example).
 """
 
 from flask_babel import gettext
-import fasttext
 import babel
 
-from searx.data import data_dir
+from searx.utils import detect_language
 from searx.languages import language_codes
 
-# Monkey patch: prevent fasttext from showing a (useless) warning when loading a
-# model.
-fasttext.FastText.eprint = lambda x: None
 
 name = gettext('Autodetect search language')
 description = gettext('Automatically detect the query search language and switch to it.')
 preference_section = 'general'
 default_on = False
 
-lang_model: fasttext.FastText._FastText = None
-"""fasttext model to predict laguage of a search term"""
-
 supported_langs = set()
 """Languages supported by most searxng engines (:py:obj:`searx.languages.language_codes`)."""
 
 
-def get_model():
-    # lazy load, in order to to save memory
-    global lang_model  # pylint: disable=global-statement
-    if lang_model is None:
-        lang_model = fasttext.load_model(str(data_dir / 'lid.176.ftz'))
-    return lang_model
-
-
 def pre_search(request, search):  # pylint: disable=unused-argument
-    prediction = get_model().predict(search.search_query.query, k=1, threshold=0.3)
-    if prediction:
-        lang = prediction[0][0].split('__label__')[1]
-        if lang in supported_langs:
-            search.search_query.lang = lang
-            try:
-                search.search_query.locale = babel.Locale.parse(lang)
-            except babel.core.UnknownLocaleError:
-                pass
+    lang = detect_language(search.search_query.query, min_probability=0)
+    if lang in supported_langs:
+        search.search_query.lang = lang
+        try:
+            search.search_query.locale = babel.Locale.parse(lang)
+        except babel.core.UnknownLocaleError:
+            pass
     return True
 
 
diff --git a/searx/search/checker/impl.py b/searx/search/checker/impl.py
index 2a30d59a7..37f145e1e 100644
--- a/searx/search/checker/impl.py
+++ b/searx/search/checker/impl.py
@@ -11,15 +11,13 @@ from urllib.parse import urlparse
 
 import re
 import httpx
-import fasttext
 
 from searx import network, logger
-from searx.utils import gen_useragent
+from searx.utils import gen_useragent, detect_language
 from searx.results import ResultContainer
 from searx.search.models import SearchQuery, EngineRef
 from searx.search.processors import EngineProcessor
 from searx.metrics import counter_inc
-from searx.data import data_dir
 
 
 logger = logger.getChild('searx.search.checker')
@@ -36,17 +34,6 @@ HTML_TAGS = [
     # fmt: on
 ]
 
-fasttext.FastText.eprint = lambda x: None
-
-lang_model: fasttext.FastText._FastText = None
-
-
-def get_model():
-    global lang_model  # pylint: disable=global-statement
-    if lang_model is None:
-        lang_model = fasttext.load_model(str(data_dir / 'lid.176.ftz'))
-    return lang_model
-
 
 def get_check_no_html():
     rep = ['<' + tag + '[^\>]*>' for tag in HTML_TAGS]
@@ -219,12 +206,8 @@ class ResultContainerTests:
         self.test_results.add_error(self.test_name, message, *args, '(' + sqstr + ')')
 
     def _add_language(self, text: str) -> typing.Optional[str]:
-        r = get_model().predict(str(text), k=1, threshold=0.3)
-        if not r:
-            return None
-
-        if len(r) > 0 and r[1][0] > 0.95:
-            langStr = r[0][0].split('__label__')[1]
+        langStr = detect_language(text)
+        if langStr:
             self.languages.add(langStr)
             self.test_results.add_language(langStr)
         return None
diff --git a/searx/utils.py b/searx/utils.py
index effb9139a..7d2c16fe2 100644
--- a/searx/utils.py
+++ b/searx/utils.py
@@ -15,6 +15,7 @@ from os.path import splitext, join
 from random import choice
 from html.parser import HTMLParser
 from urllib.parse import urljoin, urlparse
+import fasttext
 
 from lxml import html
 from lxml.etree import ElementBase, XPath, XPathError, XPathSyntaxError, _ElementStringResult, _ElementUnicodeResult
@@ -22,7 +23,7 @@ from babel.core import get_global
 
 
 from searx import settings
-from searx.data import USER_AGENTS
+from searx.data import USER_AGENTS, data_dir
 from searx.version import VERSION_TAG
 from searx.languages import language_codes
 from searx.exceptions import SearxXPathSyntaxException, SearxEngineXPathException
@@ -50,6 +51,12 @@ _STORAGE_UNIT_VALUE: Dict[str, int] = {
 _XPATH_CACHE: Dict[str, XPath] = {}
 _LANG_TO_LC_CACHE: Dict[str, Dict[str, str]] = {}
 
+_FASTTEXT_MODEL: Optional[fasttext.FastText._FastText] = None
+"""fasttext model to predict laguage of a search term"""
+
+# Monkey patch: prevent fasttext from showing a (useless) warning when loading a model.
+fasttext.FastText.eprint = lambda x: None
+
 
 class _NotSetClass:  # pylint: disable=too-few-public-methods
     """Internal class for this module, do not create instance of this class.
@@ -621,3 +628,20 @@ def eval_xpath_getindex(elements: ElementBase, xpath_spec: XPathSpecType, index:
         # to record xpath_spec
         raise SearxEngineXPathException(xpath_spec, 'index ' + str(index) + ' not found')
     return default
+
+
+def _get_fasttext_model() -> fasttext.FastText._FastText:
+    global _FASTTEXT_MODEL  # pylint: disable=global-statement
+    if _FASTTEXT_MODEL is None:
+        _FASTTEXT_MODEL = fasttext.load_model(str(data_dir / 'lid.176.ftz'))
+    return _FASTTEXT_MODEL
+
+
+def detect_language(text: str, threshold: float =0.3, min_probability: float=0.5) -> Optional[str]:
+    """https://fasttext.cc/docs/en/language-identification.html"""
+    if not isinstance(text, str):
+        raise ValueError('text must a str')
+    r = _get_fasttext_model().predict(text.replace('\n', ' '), k=1, threshold=threshold)
+    if isinstance(r, tuple) and len(r) == 2 and len(r[0]) > 0 and len(r[1]) > 0 and r[1][0] > min_probability:
+        return r[0][0].split('__label__')[1]
+    return None
diff --git a/searxng_extra/update/update_engine_descriptions.py b/searxng_extra/update/update_engine_descriptions.py
index 22ad54d9b..bc86fba85 100755
--- a/searxng_extra/update/update_engine_descriptions.py
+++ b/searxng_extra/update/update_engine_descriptions.py
@@ -23,7 +23,7 @@ from searx.utils import extract_text, match_language
 from searx.locales import LOCALE_NAMES, locales_initialize
 from searx import searx_dir
 from searx.data import data_dir
-from searx.utils import gen_useragent
+from searx.utils import gen_useragent, detect_language
 import searx.search
 import searx.network
 
@@ -127,16 +127,6 @@ def get_wikipedia_summary(lang, pageid):
         return None
 
 
-def detect_language(text):
-    r = get_model().predict(str(text), k=1, threshold=0.3)  # pylint: disable=E1101
-    if not r:
-        return None
-
-    if len(r) > 0 and r[1][0] > 0.95:
-        return r[0][0].split('__label__')[1]
-    return None
-
-
 def get_website_description(url, lang1, lang2=None):
     headers = {
         'User-Agent': gen_useragent(),
diff --git a/tests/unit/test_utils.py b/tests/unit/test_utils.py
index 8ac7db479..6f51f1ee3 100644
--- a/tests/unit/test_utils.py
+++ b/tests/unit/test_utils.py
@@ -232,3 +232,25 @@ class TestXPathUtils(SearxTestCase):
         with self.assertRaises(SearxEngineXPathException) as context:
             utils.eval_xpath_getindex(doc, 'count(//i)', 1)
         self.assertEqual(context.exception.message, 'the result is not a list')
+
+    def test_detect_language(self):
+        # make sure new line are not an issue
+        # fasttext.predict('') does not accept new line.
+        l = utils.detect_language('The quick brown fox jumps over\nthe lazy dog')
+        self.assertEqual(l, 'en')
+
+        l = utils.detect_language('いろはにほへと ちりぬるを わかよたれそ つねならむ うゐのおくやま けふこえて あさきゆめみし ゑひもせす')
+        self.assertEqual(l, 'ja')
+
+        l = utils.detect_language('Pijamalı hasta yağız şoföre çabucak güvendi.')
+        self.assertEqual(l, 'tr')
+
+        l = utils.detect_language('')
+        self.assertIsNone(l)
+
+        # mix languages --> None
+        l = utils.detect_language('The いろはにほへと Pijamalı')
+        self.assertIsNone(l)
+
+        with self.assertRaises(ValueError):
+            utils.detect_language(None)
```

**Comment by ArtikusHG** [comment]
Thanks for the feedback, I also wanted to implement detect_language in a separate place, just wasn't sure where. I will go back to this in a few days after I finish my finals.

**Comment by HLFH** [comment]
@ArtikusHG Your finals? You mean France 🇫🇷  vs Argentina 🇦🇷. Youhooou! See you on Sunday then.

**Comment by ArtikusHG** [comment]
Sorry for messing up the whole thing at first, everything should be fine now

File: searxng_extra/update/update_engine_descriptions.py [hunk 3]
```diff
@@ -80,6 +79,17 @@
...
+
+lang_model: fasttext.FastText._FastText = None
+
+
+def get_model():
+    global lang_model  # pylint: disable=global-statement
+    if lang_model is None:
+        lang_model = fasttext.load_model(str(data_dir / 'lid.176.ftz'))
+    return lang_model
+
```
Comment (dalf) [inline searxng_extra/update/update_engine_descriptions.py]:
```suggestion
```

File: searxng_extra/update/update_engine_descriptions.py [hunk 4]
```diff
@@ -15,16 +15,15 @@
...
 
-from langdetect import detect_langs
-from langdetect.lang_detect_exception import LangDetectException
-
 from searx.engines import wikidata, set_loggers
 from searx.utils import extract_text, match_language
 from searx.locales import LOCALE_NAMES, locales_initialize
 from searx import searx_dir
-from searx.utils import gen_useragent
+from searx.data import data_dir
```
Comment (dalf) [inline searxng_extra/update/update_engine_descriptions.py]:
```suggestion
```

File: searxng_extra/update/update_engine_descriptions.py [hunk 5]
```diff
@@ -15,16 +15,15 @@
 from urllib.parse import urlparse
 from os.path import join
 
+import fasttext
```
Comment (dalf) [inline searxng_extra/update/update_engine_descriptions.py]:
```suggestion
```

**Review by dalf — COMMENTED** [review]
Erf sorry, in the patch there is a left over in update_engine_descriptions.py

Also, can you merge the commits? If you have issue to do it, just let me now, we can do it.

After that we can merge the PR

**Comment by ArtikusHG** [comment]
> Also, can you merge the commits?

Sorry, I am still learning git, I tried several times to merge all commits into one but I keep getting errors, if it's not hard for you I'd rather have you do it than me mess it up :p

EDIT: nevermind, it seems I figured it out

**Comment by dalf** [comment]
Side note : you can use `git pull --rebase origin master` instead of `git merge` : it takes the new master branch and apply your commits on top of it. With `git pull --rebase ...` there is no `Merge branch` commit.

**Comment by ArtikusHG** [comment]
Oh nice, didn't know that. Will use that in the future. Is there anything more I should do?

**Comment by dalf** [comment]
All good, thank you!

**Comment by return42** [comment]
> Sorry, I am still learning git ..

@ArtikusHG the readiness for learning is always a good attitude :+1: .. if we can help you with it just ask.  May I can give you some hints at hand ..

Git is a tool for practicing collaboration in [Distributed Workflows](https://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows), a huge hub of git topics is the 

- [git-scm documentation](https://git-scm.com/doc)

The processes (aka collaboration workflows) behind it are described quite well in a paper from Atlassian.

- [Comparing Git Workflows: What You Should Know](https://www.atlassian.com/git/tutorials/comparing-workflows)

When you open a PR here on github the branch in your fork is (by default) "writeable" for the SearXNG maintainers --> by opening a PR your (feature) branch becomes a part of the SearXNG project until the PR is merged --> the PR workflow here on github is (more or less) a 

- [Git Feature Branch Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow)

> A feature branching model is a great tool to promote collaboration within a team environment.  Utilizing [git rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase) during the review and merge stages of a feature branch will create enforce a cohesive Git history of feature merges.

- [Git Branching - Rebasing](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)

**Pull request merged**